feat(components): [input-tag] add `collapse-*` prop

eighty-cents opened the pull request:
##What has this PR done?
When used as a search condition, input-tag causes the input box to expand when multiple tag contents are entered, affecting the page layout and making it unsightly. Therefore, using the collapseTag method is more appropriate.

##Screenshot/Screen Recording
<img width="1920" height="919" alt="image" src="https://github.com/user-attachments/assets/59323157-593c-45b3-b859-897fe579b9a3" />

https://github.com/user-attachments/assets/53c1f318-4a78-4471-a961-6d8d6307b1d9

Please make sure these boxes are checked before submitting your PR, thank you!

- [ ] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [ ] Make sure you are merging your commits to `dev` branch.
- [ ] Add some descriptions and refer to relative issues for your PR.

eighty-cents committed:
feat(components): Add input box label folding function

- Add collapseTags, collapseTagsTooltip, and maxCollapseTags attributes
- Implement the function of tag folding and mouse hovering to display complete content
- Update documentation and sample code

btea commented:
This feature looks good. WDYT? @tolking

eighty-cents committed:
Merge branch 'dev' into feat/input-tag-collapse

eighty-cents committed:
feat(components): optimize the judgment before out of focus

-New beforeBlur method
-Add input tag list class name to improve the style and layout of tag list
-Add label text wrapping to optimize the display of long text

eighty-cents committed:
Merge branch 'dev' into feat/input-tag-collapse

eighty-cents committed:
feat(docs): 添加标签折叠功能

- 新增 collapse-tags 属性，用于在多选时将标签折叠为文本
- 新增 collapse-tags-tooltip 属性，用于在鼠标悬停时显示折叠标签的完整内容
- 新增 max-collapse-tags 属性，用于设置折叠标签的最大数量

eighty-cents commented:
> I understand why I need these attributes, we can merge this PR.
> 
> but before that
> 
> 1. need to improve the relevant attribute documents
> 2. refer to ElSelect to complete the `beforeBlur` configuration of `useFocusController`

1. Related attribute document has been modified
2. The beforeBlur configuration has been added

tolking commented on packages/components/input-tag/src/input-tag.vue:
```diff
@@ -10,25 +10,108 @@
     </div>
     <div :class="innerKls">
+      <template v-for="(item, index) in modelValue" :key="index">
+        <el-tag
+          v-if="collapseTags && index < maxCollapseTags"
+          :size="tagSize"
+          :closable="closable"
+          :type="tagType"
+          :effect="tagEffect"
+          :draggable="closable && draggable"
+          disable-transitions
+          @close="handleRemoveTag(index)"
+          @dragstart="(event: DragEvent) => handleDragStart(event, index)"
+          @dragover="(event: DragEvent) => handleDragOver(event, index)"
+          @dragend="handleDragEnd"
+          @drop.stop
+        >
+          <slot name="tag" :value="item" :index="index">
+            {{ item }}
+          </slot>
+        </el-tag>
+        <el-tag
+          v-if="!collapseTags"
+          :size="tagSize"
+          :closable="closable"
+          :type="tagType"
+          :effect="tagEffect"
+          :draggable="closable && draggable"
+          disable-transitions
+          @close="handleRemoveTag(index)"
+          @dragstart="(event: DragEvent) => handleDragStart(event, index)"
+          @dragover="(event: DragEvent) => handleDragOver(event, index)"
+          @dragend="handleDragEnd"
+          @drop.stop
+        >
+          <slot name="tag" :value="item" :index="index">
+            {{ item }}
+          </slot>
+        </el-tag>
+      </template>
```
IMO, It would be better to calculate showTagList like ElSelect components

tolking commented on packages/components/input-tag/src/input-tag.vue:
```diff
@@ -10,25 +10,108 @@
+        </template>
+        <template #content>
+          <div :class="ns.e('input-tag-list')">
+            <div
+              v-for="(item, index) in modelValue"
+              :key="item"
+              :class="ns.e('input-tag-item')"
+            >
+              <el-tag
+                v-if="index >= maxCollapseTags"
+                :size="tagSize"
+                :closable="closable"
+                :type="tagType"
+                :effect="tagEffect"
+                disable-transitions
+                @close="handleRemoveTag(index)"
+              >
+                <span :class="ns.e('tags-text')">
+                  <slot name="label" :value="item" :index="index">
+                    {{ item }}
+                  </slot>
+                </span>
+              </el-tag>
+            </div>
```
IMO, It would be better to calculate collapseTagList like ElSelect components

It seems that unnecessary nested node `input-tag-item`, `tags-text` can be avoided here

tolking commented on docs/en-US/component/input-tag.md:
```diff
@@ -37,6 +37,16 @@ input-tag/max
 
 :::
 
+## Collapse Tags
```
```suggestion
## Collapse Tags ^(2.10.8)
```

tolking commented on docs/en-US/component/input-tag.md:
```diff
@@ -111,29 +121,32 @@ input-tag/prefix-suffix
+| draggable             | whether tags can be dragged                                                                                    | ^[boolean]                                                  | false   |
+| delimiter ^(2.9.9)    | add a tag when a delimiter is matched                                                                          | ^[string] / ^[regex]                                        | —       |
+| size                  | input box size                                                                                                 | ^[enum]`'large' \| 'default' \| 'small'`                    | —       |
+| collapse-tags         | whether to collapse tags to a text when multiple selecting                                                     | ^[boolean]                                                  | false   |
+| collapse-tags-tooltip | whether show all selected tags when mouse hover text of collapse-tags. To use this, collapse-tags must be true | ^[boolean]                                                  | false   |
```
```suggestion
| collapse-tags ^(2.10.8)         | whether to collapse tags to a text when multiple selecting                                                     | ^[boolean]                                                  | false   |
| collapse-tags-tooltip ^(2.10.8) | whether show all selected tags when mouse hover text of collapse-tags. To use this, collapse-tags must be true | ^[boolean]                                                  | false   |
```

tolking replied:
Descriptive information also needs to be optimized

tolking commented on docs/en-US/component/input-tag.md:
```diff
@@ -111,29 +121,32 @@ input-tag/prefix-suffix
+| autofocus             | same as `autofocus` in native input                                                                            | ^[boolean]                                                  | false   |
+| id                    | same as `id` in native input                                                                                   | ^[string]                                                   | —       |
+| tabindex              | same as `tabindex` in native input                                                                             | ^[string] / ^[number]                                       | —       |
+| max-collapse-tags     | the max tags number to be shown. To use this, collapse-tags must be true                                       | ^[number]                                                   | 1       |
```
```suggestion
| max-collapse-tags ^(2.10.8)     | the max tags number to be shown. To use this, collapse-tags must be true                                       | ^[number]                                                   | 1       |
```

eighty-cents committed:
Merge branch 'dev' into feat/input-tag-collapse

eighty-cents committed:
feat(components): optimize tag display logic

- Add showTagList and collapseTagList calculation properties to control the display of labels
- Modify the label removal logic, and now remove through the label value instead of the index
- Adjust the rendering method of labels to simplify the template structure
- Removed redundant tags text class styles, and directly displayed with tag values

tolking commented on packages/components/input-tag/src/composables/use-input-tag.ts:
```diff
@@ -38,6 +40,22 @@ export function useInputTag({ props, emit, formItem }: UseInputTagOptions) {
       ? false
       : (props.modelValue?.length ?? 0) >= props.max
   })
+  const showTagList = computed(() => {
+    if (!props.collapseTags) {
+      return []
+    }
+    return props.collapseTags
+      ? props.modelValue?.slice(0, props.maxCollapseTags)
+      : props.modelValue
+  })
+  const collapseTagList = computed(() => {
+    if (!props.collapseTags) {
+      return []
+    }
+    return props.collapseTags
+      ? props.modelValue?.slice(props.maxCollapseTags)
+      : []
+  })
```
```suggestion
  const showTagList = computed(() => {
    return props.collapseTags
      ? props.modelValue?.slice(0, props.maxCollapseTags)
      : props.modelValue
  })
  const collapseTagList = computed(() => {
    return props.collapseTags
      ? props.modelValue?.slice(props.maxCollapseTags)
      : []
  })
```

tolking commented on packages/components/input-tag/src/composables/use-input-tag.ts:
```diff
@@ -94,7 +112,8 @@ export function useInputTag({ props, emit, formItem }: UseInputTagOptions) {
           event.preventDefault()
           event.stopPropagation()
-          handleRemoveTag(props.modelValue.length - 1)
+          const tag = props.modelValue[props.modelValue.length - 1]
+          handleRemoveTag(tag)
```
I don't understand why we need to change this.

tolking commented on packages/components/input-tag/src/input-tag.vue:
```diff
@@ -10,25 +10,83 @@
-        :size="tagSize"
-        :closable="closable"
-        :type="tagType"
+      <template v-for="(item, index) in showTagList" :key="index">
+        <el-tag
+          :size="tagSize"
+          :closable="closable"
+          :type="tagType"
+          :effect="tagEffect"
+          :draggable="closable && draggable"
+          disable-transitions
+          @close="handleRemoveTag(item)"
+          @dragstart="(event: DragEvent) => handleDragStart(event, index)"
+          @dragover="(event: DragEvent) => handleDragOver(event, index)"
+          @dragend="handleDragEnd"
+          @drop.stop
+        >
+          <slot name="tag" :value="item" :index="index">
+            {{ item }}
+          </slot>
+        </el-tag>
+        <el-tag
+          v-if="!collapseTags"
+          :size="tagSize"
+          :closable="closable"
+          :type="tagType"
+          :effect="tagEffect"
+          :draggable="closable && draggable"
+          disable-transitions
+          @close="handleRemoveTag(item)"
+          @dragstart="(event: DragEvent) => handleDragStart(event, index)"
+          @dragover="(event: DragEvent) => handleDragOver(event, index)"
+          @dragend="handleDragEnd"
+          @drop.stop
+        >
+          <slot name="tag" :value="item" :index="index">
+            {{ item }}
+          </slot>
+        </el-tag>
+      </template>
```
If the collapseTags is false, doesn't it render the same label?

eighty-cents committed:
fix(components): fix the same label

- When collapseTags is false, the repair will render the same tags

eighty-cents committed:
refactor(components): optimize tags

- Modify the logic of fallback label deletion, and use the index to directly remove the last label
- Reconstruct the tag rendering method and simplify the template structure
- Remove unused CSS classes

eighty-cents committed:
refactor(components): optimize label rendering logic

- Replace the template tag in the template with the el tag component
- Simplify the rendering logic of labels and improve the readability and performance of components

eighty-cents committed:
refactor(components): optimize key binding of tag list

- Change the key of the tag component from item to index

eighty-cents committed:
Merge branch 'dev' into feat/input-tag-collapse

eighty-cents committed:
docs(components): Modify the description of the collapse tags attribute

- Optimize the description of collapse tags attribute

eighty-cents committed:
Merge branch 'dev' into feat/input-tag-collapse

eighty-cents committed:
feat(components): add custom slots for labels

- Add a slot named "tag" in the el tag component

tolking reviewed (approved):
Overall, there doesn't seem to be any problem.

The rest is that the content of the document needs to be optimized.

In addition, it would be best if you could add some test cases.

btea commented:
The version number corresponding to the newly added `prop` is not yet added in the documentation.

eighty-cents committed:
Merge branch 'dev' into feat/input-tag-collapse

eighty-cents committed:
test(components): add a test case for the collapseTags function

- Test the basic function of the collapseTags attribute
- est the hover effect of the collapseTagsTooltip attribute
- Test the restriction logic of the maxCollapseTags attribute

eighty-cents committed:
Update docs/en-US/component/input-tag.md

Co-authored-by: qiang <qw13131wang@gmail.com>

eighty-cents committed:
docs(components): update documents

- New features introduced in version 2.11.0 have been added to the document

eighty-cents committed:
Merge branch 'feat/input-tag-collapse' of https://github.com/eighty-cents/element-plus into feat/input-tag-collapse

eighty-cents commented:
> Overall, there doesn't seem to be any problem.
> 
> The rest is that the content of the document needs to be optimized.
> 
> In addition, it would be best if you could add some test cases.

Unit test added @tolking

eighty-cents commented:
> The version number corresponding to the newly added `prop` is not yet added in the documentation.

Corresponding version number has been added @btea

eighty-cents committed:
test(components): remove comments from unit tests

- Removed unnecessary comments in test code
